rudder-web/src/main/scala/com/normation/rudder/web/rest/technique/TechniqueAPIService6.scala

```diff
@@ -0,0 +1,137 @@
+import com.normation.rudder.web.services.DirectiveEditorService
+import com.normation.rudder.web.rest.RestExtractorService
+import com.normation.utils.Control._
+import com.normation.utils.StringUuidGenerator
+import net.liftweb.common._
+import net.liftweb.http.Req
+import net.liftweb.json.JArray
+import net.liftweb.json.JValue
+import net.liftweb.json.JsonDSL._
+import com.normation.rudder.web.rest.RestDataSerializer
+import com.normation.rudder.domain.workflows.ChangeRequestId
+import com.normation.cfclerk.domain.TechniqueId
+import com.normation.cfclerk.services.TechniqueRepository
+import com.normation.utils.Control._
+import scala.collection.SortedMap
+import net.liftweb.json.JsonAST.JValue
+
+case class TechniqueAPIService6 (
+    readDirective        : RoDirectiveRepository
+  , restDataSerializer   : RestDataSerializer
+  , techniqueRepository  : TechniqueRepository
+  ) extends Loggable {
+
+  def serialize(technique : Technique, directive:Directive) = restDataSerializer.serializeDirective(technique, directive, None)
+
+  def listTechniques : Box[JValue] = {
+    for {
+      lib <- readDirective.getFullDirectiveLibrary()
+      activeTechniques = lib.allActiveTechniques.values.toSeq
+
+      serialized = activeTechniques.map(restDataSerializer.serializeTechnique)
+    } yield {
+      serialized
+    }
+  }
+
+  def listDirectives (techniqueName : TechniqueName, wantedVersions: Option[List[TechniqueVersion]])  : Box[JValue] = {
+    def serializeDirectives (directives : Seq[Directive], techniques : SortedMap[TechniqueVersion, Technique]) = {
+     for {
+       directive <- directives
+       if ( wantedVersions match {
+         case None => true
+         case Some(versions) => versions.contains(directive.techniqueVersion)
+       })
+     } yield {
+      techniques.get(directive.techniqueVersion) match {
+        case None => Failure(s"Version ${directive.techniqueVersion} of Technique '${techniqueName.value}' does not exists, but is used by Directive '${directive.id.value}'")
+        case Some(technique) => Full(serialize(technique,directive))
+      }
+     }
+    }
+
+    for {
+      lib <- readDirective.getFullDirectiveLibrary()
+      activeTech <- Box(lib.allActiveTechniques.values.find { _.techniqueName == techniqueName }) ?~! s"Technique '${techniqueName.value}' does not exists"
+
+      // Check if version we want exists in technique library
+      _ <- wantedVersions match {
+        case Some(versions) =>
+          sequence(versions) { version =>
+            if  (activeTech.techniques.keySet.contains(version)) {
+              Full("ok")
+            } else {
+              Failure(s"Version '${version}' of Technique '${techniqueName.value}' does not exists" )
+            }
+          }
+        case None => Full("ok")
+      }
+      serializedDirectives <- boxSequence(serializeDirectives(activeTech.directives,activeTech.techniques))
+    } yield {
+      serializedDirectives.toList
+    }
+  }
```

Comment:
I would love to see all these little semantically equivalent things aligned. Much easier to parse.